clang/test/CodeGenSYCL/device_has.cpp

```diff
@@ -0,0 +1,52 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -triple spir64-unknown-unknown -disable-llvm-passes -emit-llvm %s -o - | FileCheck %s
+
+// Tests for IR of device_has(aspect, ...) attribute
+#include "sycl.hpp"
+
+using namespace cl::sycl;
+queue q;
+
+// CHECK: define dso_local spir_func void @{{.*}}func1{{.*}} !intel_declared_aspects ![[ASPECTS1:[0-9]+]] {
+[[sycl::device_has(cl::sycl::aspect::cpu)]] void func1() {}
+
+// CHECK: define dso_local spir_func void @{{.*}}func2{{.*}} !intel_declared_aspects ![[ASPECTS2:[0-9]+]] {
+[[sycl::device_has(cl::sycl::aspect::fp16, cl::sycl::aspect::gpu)]] void func2() {}
+
+// CHECK: define dso_local spir_func void @{{.*}}func3{{.*}} !intel_declared_aspects ![[EMPTYASPECTS:[0-9]+]] {
+[[sycl::device_has()]] void func3() {}
+
+// CHECK: define linkonce_odr spir_func void @{{.*}}func4{{.*}} !intel_declared_aspects ![[ASPECTS3:[0-9]+]] {
+template <cl::sycl::aspect Aspect>
+[[sycl::device_has(Aspect)]] void func4() {}
+
+// CHECK: define dso_local spir_func void @{{.*}}func5{{.*}} !intel_declared_aspects ![[ASPECTS1]] {
+[[sycl::device_has(cl::sycl::aspect::cpu)]] void func5();
+void func5() {}
+
+[[sycl::device_has(cl::sycl::aspect::cpu)]] void func6();
+// CHECK: define dso_local spir_func void @{{.*}}func6{{.*}} !intel_declared_aspects ![[ASPECTS1]] {
+void func6() {}
+
+class KernelFunctor {
+public:
+  void operator()() const {
+    func1();
+    func2();
+    func3();
+    func4<cl::sycl::aspect::host>();
+    func5();
+    func6();
+  }
+};
+
+void foo() {
+  q.submit([&](handler &h) {
+    KernelFunctor f1;
+    h.single_task<class kernel_name_1>(f1);
+  });
+}
```

Comment:
Should there be some tests in here for the case when the parameter to `sycl::device_has()` is a `constexpr` expression, rather than a literal value?  For example, something like:

```
constexpr cl::sycl::aspect get_aspect() {return cl::sycl::aspect::fp16;}
[[sycl::device_has(get_aspect())]] void func6();
```